src/core/eme/handle_session_events.ts

```diff
@@ -0,0 +1,208 @@
+  value : {
+    session : IMediaKeySession|MediaKeySession;
+    license: ILicense;
+  };
+}
+
+/**
+ * listen to "message" events from session containing a challenge
+ * blob and map them to licenses using the getLicense method from
+ * selected keySystem.
+ * @param {MediaKeySession} session
+ * @param {Object} keySystem
+ * @param {Subject} errorStream
+ * @returns {Observable}
+ */
+export default function handleSessionEvents(
+  session: IMediaKeySession|MediaKeySession,
+  keySystem: IKeySystemOption,
+  errorStream: Subject<Error|CustomError>
+) : Observable<IMediaKeySessionHandledEvents> {
+  log.debug("eme: handle message events", session);
+
+  /**
+   * @param {Error|Object} error
+   * @param {Boolean} fatal
+   * @returns {Error|Object}
+   */
+  function licenseErrorSelector(
+    error: CustomError|Error,
+    fatal: boolean
+  ) : CustomError|Error {
+    if (isKnownError(error)) {
+      if (error.type === ErrorTypes.ENCRYPTED_MEDIA_ERROR) {
+        error.fatal = fatal;
+        return error;
+      }
+    }
+    return new EncryptedMediaError("KEY_LOAD_ERROR", error, fatal);
+  }
+
+  const getLicenseRetryOptions = {
+    totalRetry: 2,
+    retryDelay: 200,
+    errorSelector: (error: CustomError|Error) => licenseErrorSelector(error, true),
+    onRetry: (
+      error: CustomError|Error) => errorStream.next(licenseErrorSelector(error, false)
+    ),
+  };
+
+  const keyErrors: Observable<never> = onKeyError$(session)
+    .map((error) => {
+      throw new EncryptedMediaError("KEY_ERROR", error, true);
+    });
+
+  const keyStatusesChanges : Observable<IMediaKeySessionEvents> =
+    onKeyStatusesChange$(session).mergeMap((keyStatusesEvent: Event) => {
+      log.debug(
+        "eme: keystatuseschange event",
+        session,
+        keyStatusesEvent
+      );
+
+      // find out possible errors associated with this event
+      session.keyStatuses.forEach((keyStatus : string, keyId : string) => {
+        // Hack present because the order of the arguments has changed in spec
+        // and is not the same between some versions of Edge and Chrome.
+        if (KEY_STATUS_ERRORS[keyId]) {
+          throw new EncryptedMediaError("KEY_STATUS_CHANGE_ERROR", keyId, true);
+        } else if (KEY_STATUS_ERRORS[keyStatus]) {
+          throw new EncryptedMediaError("KEY_STATUS_CHANGE_ERROR", keyStatus, true);
+        }
+      });
+
+      const handledKeyStatusesChange$ = tryCatch(() => {
+        return keySystem && keySystem.onKeyStatusesChange ?
+          castToObservable(
+            keySystem.onKeyStatusesChange(keyStatusesEvent, session)
+          ) as Observable<TypedArray|ArrayBuffer> : Observable.empty<never>();
+      });
+
+      return handledKeyStatusesChange$
+        .catch((error: Error) => {
+          throw new EncryptedMediaError("KEY_STATUS_CHANGE_ERROR", error, true);
+        })
+        .map((licenseObject) => {
+          return {
+            type: "key-status-change" as "key-status-change",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=222&issues=AWNklCZBXOBV_YrhYTpZ&open=AWNklCZBXOBV_YrhYTpZ)

[ISSUE_KEY]: AWNklCZBXOBV_YrhYTpZ